dns/module.go

```diff
@@ -0,0 +1,342 @@
+	}
+
+	return &ModuleInstance{
+		vu:        vu,
+		dnsClient: NewDNSClient(),
+		metrics:   instanceMetrics,
+	}
+}
+
+// Exports returns the module exports, that will be available in the runtime.
+func (mi *ModuleInstance) Exports() modules.Exports {
+	return modules.Exports{Named: map[string]interface{}{
+		"resolve": mi.Resolve,
+		"lookup":  mi.Lookup,
+	}}
+}
+
+// Resolve resolves a domain name to an IP address.
+// func (mi *ModuleInstance) Resolve(query, recordType sobek.Value, resolveDNSOptions *sobek.Object) *sobek.Promise {
+func (mi *ModuleInstance) Resolve(query, recordType, nameserverAddr sobek.Value) *sobek.Promise {
+	promise, resolve, reject := promises.New(mi.vu)
+
+	if mi.vu.State() == nil {
+		reject(errors.New("resolve can not be used in the init context"))
+		return promise
+	}
+
+	var queryStr string
+	if err := mi.vu.Runtime().ExportTo(query, &queryStr); err != nil {
+		reject(fmt.Errorf("query must be a string; got %v instead", query))
+		return promise
+	}
+
+	var recordTypeStr string
+	if err := mi.vu.Runtime().ExportTo(recordType, &recordTypeStr); err != nil {
+		reject(fmt.Errorf("recordType must be a string; got %v instead", recordType))
+		return promise
+	}
+
+	var nameserverAddrStr string
+	if err := mi.vu.Runtime().ExportTo(nameserverAddr, &nameserverAddrStr); err != nil {
+		reject(fmt.Errorf("nameserver must be a string; got %v instead", nameserverAddr))
+		return promise
+	}
+
+	// nameserver := NewNameserver(options.Nameserver.IP, options.Nameserver.Port)
+	nameserver, err := ParseNameserverAddr(nameserverAddrStr)
+	if err != nil {
+		reject(fmt.Errorf("parsing nameserver address failed: %w", err))
+		return promise
+	}
+
+	go func() {
+		// Start timer for resolution
+		resolutionStartTime := time.Now()
+
+		// Resolve the query
+		fetchedIPs, resolveErr := mi.dnsClient.Resolve(mi.vu.Context(), queryStr, recordTypeStr, nameserver)
+
+		// Stop the timer for resolution
+		sinceResolutionStart := time.Since(resolutionStartTime).Milliseconds()
+
+		// Emit the metrics, regardless of the result
+		mi.emitResolutionMetrics(
+			mi.vu.Context(),
+			sinceResolutionStart,
+			queryStr,
+			recordTypeStr,
+			nameserver,
+			resolveErr,
+		)
+
+		// Handle the resolution failure only now that we have emitted the metrics
+		if resolveErr != nil {
+			reject(resolveErr)
+			return
+		}
+
+		resolve(fetchedIPs)
+	}()
+
+	return promise
+}
+
+// Lookup resolves a domain name to an IP address using the default system nameservers.
+func (mi *ModuleInstance) Lookup(hostname sobek.Value) *sobek.Promise {
+	promise, resolve, reject := promises.New(mi.vu)
+
+	if mi.vu.State() == nil {
+		reject(fmt.Errorf("lookup can not be used in the init context"))
+		return promise
+	}
+
+	var hostnameStr string
+	if err := mi.vu.Runtime().ExportTo(hostname, &hostnameStr); err != nil {
+		reject(fmt.Errorf("hostname must be a string; got %v instead", hostname))
```

Comment:
From the context, I guess you meant type, not the value, right?

```suggestion
		reject(fmt.Errorf("hostname must be a string; got %T instead", hostname))
```